File: ABC/ABC005/D_50.cpp
// 部分点50点解法 全探索O(N^6・Q) (Q=N^2としてO(N^8))

#include <bits/stdc++.h>
using namespace std;

int main() {
    int n, q, p;
    cin >> n;
    vector<vector<int>> d(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) cin >> d[i][j];
    }
    cin >> q;
    while (cin >> p) {
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (i * j > p) break;
                for (int s = 0; s <= n - i; s++) {
                    for (int t = 0; t <= n - j; t++) {
                        int now = 0;
                        for (int k = 0; k < i; k++) {
                            for (int l = 0; l < j; l++) {
                                now += d[s + k][t + l];
                            }
                        }
                        ans = max(ans, now);
                    }
                }
            }
        }
        cout << ans << '\n';
    }
}

File: ABC/ABC005/D_50.py
n = int(input())
d = [list(map(int, input().split())) for _ in range(n)]
for _ in range(int(input())):
    p = int(input())
    ans = 0
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if i * j > p:
                break
            for s in range(n - i + 1):
                for t in range(n - j + 1):
                    now = 0
                    for k in range(i):
                        for l in range(j):
                            now += d[s + k][t + l]
                    ans = max(ans, now)
    print(ans)